src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SuppressGCTransitionAttribute.cs

```diff
@@ -0,0 +1,54 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// An attribute used to indicate a GC transition should be skipped when making an unmanaged function call.
+    /// </summary>
+    /// <example>
+    /// Example of a valid use case. The Win32 `GetTickCount()` function is a small performance related function
+    /// that reads some global memory and returns the value. In this case, the GC transition overhead is significantly
+    /// more than the memory read.
+    /// <code>
+    /// using System;
+    /// using System.Runtime.InteropServices;
+    /// class Program
+    /// {
+    ///     [DllImport("Kernel32")]
+    ///     [SuppressGCTransition]
+    ///     static extern int GetTickCount();
+    ///     static void Main()
+    ///     {
+    ///         Console.WriteLine($"{GetTickCount()}");
+    ///     }
+    /// }
+    /// </code>
+    /// </example>
+    /// <remarks>
+    /// This attribute is ignored if applied to a method without the <see cref="System.Runtime.InteropServices.DllImportAttribute"/>.
+    ///
+    /// Supressing the GC transition is an advanced scenario and should not be done without fully understanding
+    /// potential consequences. Examples of cases where this attribute is not appropriate:
+    ///   * Native function calls back into the runtime (e.g. Reverse P/Invoke).
+    ///   * Native function performs some type of I/O.
+    ///   * Native function executes for a non-trivial amount of time.
+    ///   * Native function requires extensive object manipulation during marshalling.
+    ///
+    ///  Cases where the attribute can be considered:
+    ///   * Native function consists of a trivial number of instructions.
+    ///   * Native function performs a task that takes less time to execute than a GC transition.
+    ///
+    /// Consequences of invalid uses of this attribute:
+    ///   * GC starvation.
+    ///   * Immediate runtime termination.
+    /// </remarks>
```

Comment:
One of the themes I'd like to see us embrace for .NET 5 are analyzers.  The platform's evolved heavily since we last had any kind of static analysis push, and with new sharp edges like this, it'd be helpful if analyzers could help to both flag places where they're used incorrectly and places where they could be used for benefit.  Since there's very little code to analyze here, it might be challenging, but I wonder if we might be able to have some heuristics, e.g. flag methods that take certain argument types (delegates), flag methods with marshaling attributes or types that suggest complicated marshaling, names that include I/O sounding operations or even known Win32 and POSIX methods, etc.

Reply:
I have captured this in the list of issues to file if this PR is merged in. See this PR description for the list.